ChangeLog.d/8710.txt

```diff
@@ -0,0 +1,3 @@
+Features
+   * PK provides support for opaque keys as soon as MBEDTLS_CRYPTO_CLIENT is
+     enabled, regardless of the status of MBEDTLS_USE_PSA_CRYPTO.
```

Comment:
Blocker: as of 4b2700e14934b815e73f9ef39a08705b732b7acc, the code in Mbed TLS that consumes pk keys does not correctly deal with the `MBEDTLS_PK_OPAQUE` case when `MBEDTLS_USE_PSA_CRYPTO` is disabled. I have not yet fully reviewed this aspect, but it's something we need to do.

At least one case that I've found while working on expanding testing: if you configure a TLS server for static ECDH with an opaque key, it crashes due to a null pointer dereference when it tries to use the key for ECDH. Found with the `Opaque key for server authentication: ECDH-` test of `ssl-opt.sh` after expanding the test programs. Backtrace:
```
0 in mbedtls_ecdh_get_params of source/library/ecdh.c:448
1 in ssl_get_ecdh_params_from_cert of source/library/ssl_tls12_server.c:2751
2 in ssl_write_server_key_exchange of source/library/ssl_tls12_server.c:3264
3 in mbedtls_ssl_handshake_server_step of source/library/ssl_tls12_server.c:4356
4 in mbedtls_ssl_handshake_step of source/library/ssl_tls.c:4022
5 in mbedtls_ssl_handshake of source/library/ssl_tls.c:4069
6 in main of source/programs/ssl/ssl_server2.c:3463
```
The code for call 1 is
```
    if ((ret = mbedtls_ecdh_get_params(&ssl->handshake->ecdh_ctx,
                                       mbedtls_pk_ec_ro(*mbedtls_ssl_own_key(ssl)),
                                       MBEDTLS_ECDH_OURS)) != 0) {
```
and `mbedtls_pk_ec_ro` returns `NULL` when the key is opaque.

Reply:
Having dug further, I believe the only risk of a security hole or crash (as opposed to merely not functioning) are where the code calls one of the pk unsafe functions (`mbedtls_pk_rsa`, `mbedtls_pk_ec{,_ro,_rw}`). I've checked and I'm happy to report that the static ECDH case mentioned above is the only one. Making it work looks difficult, but I'm ok with releasing 3.6.0 with this as a documented limitation that causes an internal error at runtime.

Reply:
https://github.com/Mbed-TLS/mbedtls/pull/8910 has passed the CI, so I recommend including it here. I believe it fully resolves this comment.

Reply:
> Making it work looks difficult

My initial reaction when seeing #8905 was "this might be more difficult than you think", and then "is this really worth it?" Apparently for PK opaque you managed to make it work (I'm assuming, since 8910 is now passing CI). The PSK case may well turn out to be harder (I'm mostly worried about the mixed-PSK key exchanges RSA-PSK, (EC)DHE-PSK).

But I stand by my second point: why would anybody who cares about PSA in TLS not want to enable `USE_PSA_CRYPTO`? As discussed on Slack, that makes no sense to me.

> but I'm ok with releasing 3.6.0 with this as a documented limitation that causes an internal error at runtime.

That no longer needs to be considered since you made it work, but if it were I'd have argued against it. Having things that partially work impose an extra cognitive burden on our users and ourselves, not to mention making documenting and testing harder. And again, for very little gain in exchange: does anyone really want to use PSA crypto in TLS and not enable `USE_PSA_CRYPTO`? Why would anyone want that?

Currently the semantics of `USE_PSA_CRYPTO` is more complicated than it should be, for our user's sake as well as ours. On one hand, I can see how making the extra APIs like `setup_opaque` depend only on `_C` (or `_CLIENT` unofficially) could be a step towards making it simpler. On the other hand, from an implementor's perspective, I think it makes thinks more complicated, specifically answering "should this crypto operation in TLS 1.2 or X.509 or PK be done with PSA APIs or legacy APIs?".

Currently the answer is a simple "if and only if `USE_PSA_CRYPTO` is enabled" (if we disregard restartable ECC for argument's sake - but even then, the decision is made at compile time). If we generalize `pk_setup_opaque()` and `ssl_xxx_psk_opaque()` to be present with only `_C` I'm afraid the answer is going to be more complicated: "is `USE_PSA_CRYPTO` enabled? if not, is the core related to an opaque thing in a way that we should still use PSA anyway?" Not sure we want that.

(Note: writing this before I took the time to look at 8910 or investigate about 8905. Perhaps after I do I'll realise things turn out to be simpler than I think. I'm still thinking this looks like non-trivial effort and potential for bad surprises, for something of very little value.)

Reply:
> Who wants to use PSA-held keys with TLS but also that TLS use the legacy API for everything else? 

A number of people don't use `MBEDTLS_USE_PSA_CRYPTO`:

* People who use Linux distributions (or BSD ports, etc.) built with the default configuration.
* People who see that we've fixed some pretty bad bugs with `MBEDTLS_USE_PSA_CRYPTO` and are worried what else we haven't discovered yet.

Some of those people are still interested in secure element support in TLS. Admittedly, while PSA opaque keys are the most natural way to achieve this, there are other ways to do it for signature-based key exchanges (such as `MBEDTLS_SSL_ASYNC_PRIVATE`).

> especially for the PSK case with XXX-PSK key exchanges

PSK is not often used with the constraint “must work with the default configuration”, so that's less important.

> I don't think we want to rush this in 3.6.0

That I agree with. I'm at the point where I'm still hesitating how to best get the `CLIENT` extension: redo this PR and review it from scratch, or just rebase it and get the OPAQUE extension as well?

Reply:
Following our discussion on Slack, this discussion is resolved as far as I'm concerned: you're not confident that we can have something reasonable in 3.6.0, so we aren't generalizing `MBEDTLS_USE_PSA_CRYPTO` opaque.

The question is then whether to remove the USE generalization from this PR, or to make a new PR with a clean commit history that only does the C→CLIENT generalization. We've decided that a clean commit history will make the review easier.